src/helpers/check_gradient.jl

```diff
@@ -0,0 +1,119 @@
+    window=nothing,
+)
+    gradient = gradF(M, p)
+    Xn = X ./ norm(M, p, X) # normalize tangent direction
+    is_vector(M, p, gradient, error)
+    # function for the directional derivative
+    df(M, p, Y) = inner(M, p, gradient, Y)
+    #
+    T = exp10.(log_range)
+    # points p_i to evaluate our error function at
+    points = map(t -> retract(M, p, Xn, t, retraction_method), T)
+    # F(p_i)
+    costs = [F(M, pi) for pi in points]
+    # linearized
+    linearized = map(t -> F(M, p) + t * df(M, p, Xn), T)
+    L = abs.(costs .- linearized)
+    # global fit a + bx
+    x = log_range[L .> 0]
+    y = log10.(L[L .> 0])
+    (a, b) = find_best_slope_window(x, y, length(x))[1:2]
+    if isapprox(b, 2.0; atol=slope_tol)
+        plot &&
+            plot_slope(T[L .> 0], L[L .> 0]; line_base=costs[1], a=a, b=b, i=1, j=length(y))
+        (io !== nothing) && print(
+            io,
+            "You gradients slope is globally $(@sprintf("%.4f", b)), so within 2 ± $(slope_tol).\n",
+        )
+        return true
+    end
+    # otherwise
+    # find best contiguous window of length w
+    (ab, bb, ib, jb) = find_best_slope_window(log_range, log10.(L), window)
+    plot && plot_slope(T[L .> 0], L[L .> 0]; line_base=costs[1], a=ab, b=bb, i=ib, j=jb)
+    (io !== nothing) && print(
+        io,
+        "You gradient fits best on [$(T[ib]),$(T[jb])] with slope  $(@sprintf("%.4f", bb)), but globally your slope $(@sprintf("%.4f", b)) is outside of the tolerance 2 ± $(slope_tol).\n",
+    )
+    return false
+end
+
+"""
+    (a,b,i,j) = find_best_slope_window(X,Y,window=nothing; slope=2.0)
+
+Check data X,Y for the best contiguous interval with a regression line closest to `slope`.
+
+If the window is set to `nothing` (default), all window sizes `2,...,length(X)` are checked.
+You can also specify a window size or an array of window sizes.
+
+For each window size , all its translates in the data are checked.
+For all these (shifted) windows the regression line is computed (i.e. `a,b` in `a + t*b`)
+and the best line is computed.
+
+From the best line the following data is returned
+
+* `a`, b` specifying the regression line `a + tb`
+* `i`, `j` determining the window, i.e the regression line stems from data `X[i], ..., X[j]`
+"""
+function find_best_slope_window(X, Y, window=nothing; slope=2.0)
+    n = length(X)
+    a_best = 0
+    b_best = 0
+    i_best = 0
+    j_best = 0
+    for w in (window === nothing ? (2:n) : [window...])
+        for j in 1:(n - w + 1)
+            x = X[j:(j + w - 1)]
+            y = Y[j:(j + w - 1)]
+            # fit a line a + bx
+            c = cor(x, y)
+            b = std(y) / std(x) * c
+            a = mean(y) - b * mean(x)
+            # look for the best error relative to log scale interval
+            r = (maximum(x) - minimum(x))
+            if abs(b - slope) < abs(b_best - slope)
+                a_best = a
+                b_best = b
+                i_best = j
+                j_best = j + w
```

Comment:
`j_best` can be outside the range of X. You are using `ib` and `jb` in line 72 to access element of the vector, which can throw an index out of bounds error.

Reply:
It can only be, if the user specifies a too large window. So if it is `nothing` then all possible sizes are chosen, then nothing happens. Since j is chosen the right way for any window – a too large window might just yield a negative j (so i_best would cause an error).
I'll add a check for the windows.

Reply:
I have code that throws an error in https://github.com/JuliaManifolds/Manopt.jl/pull/117/files#diff-83a07696aae40baf5007ef0f950a7bad4548d0180ae80ac2c4397cb75585830bR71

```
check_gradient(TR, cost_fnc, grad_fnc, plot = true)
ERROR: BoundsError: attempt to access 101-element Vector{Float64} at index [102]
Stacktrace:
 [1] getindex
   @ ./array.jl:861 [inlined]
 [2] check_gradient(M::TangentSpaceAtPoint{ℝ, Rotations{3}, Matrix{Float64}}, F::typeof(cost_fnc), gradF::typeof(grad_fnc), p::Matrix{Float64}, X::Matrix{Float64}; plot::Bool, error::Bool, io::Nothing, limits::Tuple{Float64, Float64}, N::Int64, log_range::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}, slope_tol::Float64, window::Nothing)
   @ Main /private/tmp/check_grad_review.jl:56
 [3] top-level scope
   @ ~/prog/julia/manifold_derivations/2022-03-22_mnn_overlay.jl:119
```

Unfortunately, the code is not particularly reproducible, but I thought the root cause was this line, but I could also be wrong about this

Reply:
hm, interesting, surely a one-off-error – for example in the very beginning there were a few ±1s missing in the idices but I had hoped to have resolved them (they do not a--read in my current first test cases).

But very cool that you are already using this :) that helps finding exactly those mistakes. I will have to take a closer look soon into the indices then.

Reply:
Ha! But I took a first close look and missed to update `j_best` correctly. If you compare what we extract in `x` and `y` -- namely `j:(j+w-1)` (that is _w_ entries j, j+1, ..., j+w-1, since the first we could is basically entry zero), then `j_best` has to be `j+w-1` as well, and this only throws an error if your best window actually reaches into the last index. It's fixed in the newest commit :)

Reply:
Great :)